Add structured outputs (#78)
webmachinelearning/prompt-api: README.md

We'll likely explore more specific APIs for tool- and function-calling in the future; follow along in [issue #7](https://github.com/webmachinelearning/prompt-api/issues/7).

### Structured output or JSON output

To help with programmatic processing of language model responses, the prompt API supports structured outputs defined by a JSON schema.

```js
const session = await ai.languageModel.create();

const responseJSONSchemaObj = new AILanguageModelResponseSchema({
  "type" : "object",
  "required" : ["Rating"],
  "additionalProperties" : false,
  "properties" : {
    "Rating": {
      "type": "number"
      "minimum": 0,
      "maximum": 5
    }
  }
});

// Prompt the model and wait for the json response to come back.
const result = await session.prompt("Summarize this feedback into a rating between 0-5: "+
  "The food was delicious, service was excellent, will recommend.",
  {responseJSONSchema : responseJSONSchemaObj}
);

> console.log(result);

Reviewer: Will `result` be a JS value, or a string that needs to be `JSON.parse()`ed?

A JS value would be nicer for developers, but I can't see how to make it work with `promptStreaming()`, so maybe it should be a string?

Author: Yup would preclude streaming, and we would have two different types of return objects from the same API. Leaving it as string.

Reviewer: Can you explain this in the explainer?

Author: done

```

> The responseJSONSchema field in the prompt options passed to session.prompt can accept a JSON Schema directly as a JavaScript object. This is particularly useful for cases where the schema is not reused for other prompts. If required developers should validate the response against the schema to ensure full conformity, as some aspects of the JSON schema may not be enforced by the API.

Reviewer: ```suggestion
The `responseJSONSchema` option for `prompt()` and `promptStreaming()` can also accept a JSON schema directly as a JavaScript object. This is particularly useful for cases where the schema is not reused for other prompts.

If perfect conformance is required, developers should validate the response against the schema to ensure full conformity, as some aspects of the JSON schema may not be enforced by the API.
```

> The responseJSONSchema field in the prompt options passed to session.prompt can accept a JSON Schema directly as a JavaScript object. This is particularly useful for cases where the schema is not reused for other prompts. If required developers should validate the response against the schema to ensure full conformity, as some aspects of the JSON schema may not be enforced by the API.

Reviewer: Could we do this validation step for developers? I guess it would require adding JSON schema validation code to the implementation. But maybe it is worth it, to avoid implementation differences in how much the schema is respected between different browsers?

Reviewer: I'm a bit uncomfortable with this lack of interoperability. At the very least, I do not think we should clearly state the proposed strategy; we should be more cautious and state it as a tentative strategy that is open to stakeholder feedback.

I think the change to raise an error when the schema is not supported is good. However, it should be a `"NotSupportedError"` `DOMException`, instead of an `"AbortError"`. And, I wonder if we'll be able to achieve interoperability on what subset of the [JSON schema spec](https://json-schema.org/specification) is supported. That is what we would normally require for web APIs. This API is a bit special, so maybe we won't be able to do, but it should be our baseline. Do you think it is possible?

For this part

> API does not guarantee conformance of the output to specified schema, if required developers should validate the response against the schema to ensure full conformity, as some aspects of the JSON schema may not be enforced.

as I stated above this is uncomfortable. It seems you are implying there will be three types of features from the JSON schema spec:

- The ones that our implementation enforces
- The ones that our implementation does not enforce, but our implementation knows it does not enforce: it will raise a `"NotSupportedError"` `DOMException`
- The ones that our implementation does not enforce, but our implementation does not know it does not enforce: these will silently fail to be enforced

Is this correct? If so, do you think it would be possible to eliminate this third type?

Author: I hear in case of llguidance it already returns errors for unsupported features, so I think it is possible to eliminate 3. Even if any other library has gaps, detecting if a schema feature is not supported from an input schema later does not look like an impossible ask. Updated the explainer to say "NotSupportedError" will be thrown for all cases where the schema is not understood by the guidance library.

We'll likely explore more specific APIs for tool- and function-calling in the future; follow along in [issue #7](https://github.com/webmachinelearning/prompt-api/issues/7).

### Structured output or JSON output

To help with programmatic processing of language model responses, the prompt API supports structured outputs defined by a JSON schema.

```js
const session = await ai.languageModel.create();

> const responseJSONSchemaObj = new AILanguageModelResponseSchema({
>   "type" : "object",
>   "required" : ["Rating"],
>   "additionalProperties" : false,
>   "properties" : {
>     "Rating": {
>       "type": "number"
>       "minimum": 0,
>       "maximum": 5
>     }
>   }
> });

Reviewer: ```suggestion
const responseJSONSchemaObj = new AILanguageModelResponseSchema({
  type: "object",
  required: ["Rating"],
  additionalProperties: false,
  properties: {
    Rating: {
      type: "number",
      minimum: 0,
      maximum: 5,
    },
  },
});
```

(comma after `"number"` was missing, and the spacing around `:` was inconsistent. Removing quotes on property names is just a style choice.)

// Prompt the model and wait for the json response to come back.
const result = await session.prompt("Summarize this feedback into a rating between 0-5: "+
  "The food was delicious, service was excellent, will recommend.",
  {responseJSONSchema : responseJSONSchemaObj}
);
console.log(result);
```

The `responseJSONSchema` option for `prompt()` and `promptStreaming()` can also accept a JSON schema directly as a JavaScript object. This is particularly useful for cases where the schema is not reused for other prompts.

> While processing the JSON schema, in cases where the user agent detects unsupported schema a `"NotSupportedError"` `DOMException`, will be raised with appropriate error message. The result value returned is a string, that can be parsed with `JSON.parse()`. If the user agent detects the prompt response as not being complaint with the schema an `"AbortError"` `DOMException` will be raised.

Reviewer: ```suggestion
While processing the JSON schema, in cases where the user agent detects unsupported schema a `"NotSupportedError"` `DOMException`, will be raised with appropriate error message. The result value returned is a string, that can be parsed with `JSON.parse()`. If the user agent is unable to produce a response that is compliant with the schema, a `"SyntaxError"` `DOMException` will be raised.
```

- I rephrased this to talk about the results the web developer cares about (cannot produce the result they want), instead of the implementation strategy (try, and detect if we fail). I think this is correct, but please check my logic.
- We should not use "`AbortError`" for this, since that's used when the web developer aborts. I think "`SyntaxError`" is reasonable, but there's no perfect fit.

Reviewer 2: Lgtm

  readonly attribute float maxTemperature;
};

> [Exposed=Window]

Reviewer: ```suggestion
[Exposed=(Window,Worker)]
```
